src/array.c

```diff
@@ -399,14 +410,37 @@ const void * const*array_get_buffer(Array *ar)
  */
 enum cc_stat array_index_of(Array *ar, void *element, size_t *index)
 {
-    size_t i;
-    for (i = 0; i < ar->size; i++) {
-        if (ar->buffer[i] == element) {
-            *index = i;
-            return CC_OK;
+    if(ar->is_sorted){
+        size_t left=0, right = ar->size-1, middle;
+	int cmp;
+	bool done=false;
+	while(!done){
+            middle = (right-left)/2 + left;
+	    cmp = ar->comp(element,ar->buffer[middle]);
+	    if(cmp < 0){
+                right = middle;
+	    }else if(cmp > 0){
+                left = middle;
+	    }else if(cmp == 0){
+                done = true;
+	    }
```

Comment:
This bsearch implementation is buggy because it can never reach the highest index. Here's an illustration of the problem:

    size = 5    

    First division:

      left   = 0
      right  = 4
      middle = (4 - 0) / 2 + 0 = 2

        0   1   2   3   4
      +---+---+---+---+---+
      |   |   |   |   | x |
      +---+---+---+---+---+
                ^
                |

      left = middle

    Second division:

      left   = 2
      right  = 4
      middle = (4 - 2) / 2 + 2 = 3

        0   1   2   3   4
      +---+---+---+---+---+
      |   |   |   |   | x |
      +---+---+---+---+---+
                    ^
                    |

      left = middle

    Third division:

      left   = 3
      right  = 4
      middle = (4 - 3) / 2  + 3 = 3    // 1 is divided by 2 to give us 0 because of rounding

      This is where the algorithm gets stuck

        0   1   2   3   4
      +---+---+---+---+---+
      |   |   |   |   | x |
      +---+---+---+---+---+
                    ^
                    |

      left = middle

    ...and so on into infinity